### PR TITLE
chore(provider): simplify history inserts

### PR DESCRIPTION
**Pull request by rkrasiuk**
Extracted from https://github.com/paradigmxyz/reth/pull/3082

## Description

This PR simplifies and generalizes history insert functions, makes them reusable.

### CODE REVIEW

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -736,20 +735,66 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
...
-        let mut cursor = self.tx.cursor_read::<tables::StorageHistory>()?;
-        let last = cursor.seek_exact(StorageShardedKey::new(address, storage_key, u64::MAX))?;
-        if let Some((storage_shard_key, list)) = last {
+    fn take_shard<T>(&self, key: T::Key) -> Result<Vec<u64>>
+    where
+        T: Table<Value = BlockNumberList>,
+    {
+        let mut cursor = self.tx.cursor_read::<T>()?;
+        let last = cursor.seek_exact(key)?;
+        if let Some((shard_key, list)) = last {
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:746]: nit
```suggestion
        let shard = cursor.seek_exact(key)?;
        if let Some((shard_key, list)) = shard {
```

File: crates/storage/provider/src/providers/database/provider.rs [hunk 2]
```diff
@@ -736,20 +735,66 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
...
     }
 
+    /// Insert history index to the database.
+    ///
+    /// For each updated partial key, this function removes the last shard from
+    /// the database (if any), appends the new indices to it, chunks the resulting integer list and
+    /// inserts the new shards back into the database.
+    ///
+    /// This function is used by history indexing stages.
+    fn insert_history_index<P, T>(
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:762]: IIUC this function is expected to be called in append way, so we can't insert indices that are less than the last inserted index. So just a small nit on naming for clarity
```suggestion
    fn append_history_index<P, T>(
```

**Review by shekhirin — COMMENTED** [review]
LGTM, small nits

File: crates/storage/provider/src/providers/database/provider.rs [hunk 3]
```diff
@@ -736,20 +735,66 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
         Ok(())
     }
 
-    /// Load last shard and check if it is full and remove if it is not. If list is empty, last
+    /// Load shard and check if it is full and remove if it is not. If list is empty, last
```
Comment (shekhirin) [inline crates/storage/provider/src/providers/database/provider.rs:738]: I'm not very familiar with shards logic, but do we still check if the shard is full in this function?

**Review by rakita — APPROVED** [review]
I like this a lot! Few nits from @shekhirin

File: crates/storage/provider/src/providers/database/provider.rs (hunk 3 above)
Comment (shekhirin) [reply crates/storage/provider/src/providers/database/provider.rs:738]: Yep, makes sense. So something like this?
```suggestion
    /// Load shard and remove it. If list is empty, last
```

**Review by shekhirin — APPROVED** [review]

**Pull request merged**